### PR TITLE
User defined metadata for uploaded S3 products

### LINKED ISSUE #50: As a user, I want to include a MD5 checksum in the the user-defined object metadata being sent in the upload payload
### Checked for duplicates

No - I haven't checked

### 🧑‍🔬 User Persona(s)

Archivist

### 💪 Motivation

...so that I can include a checksum with the files being uploaded to ensure data integrity as the files flow through the system

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** a file to be uploaded to S3
**When I perform** data upload manager execution on that file
**Then I expect** data upload manager to generate a checksum and add to the object metadata for the S3 object

### ⚙️ Engineering Details

https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html

### LINKED ISSUE #87: As a user, I want to include the modification datetime in the the user-defined object metadata being sent in the upload payload
### Checked for duplicates

Yes - I've already checked

### 🧑‍🔬 User Persona(s)

Archivist

### 💪 Motivation

...so that I can match the modification datetime from the source system where the data is being copied.

### 📖 Additional Details

_No response_

### Acceptance Criteria

**Given** 
**When I perform** 
**Then I expect** 

### ⚙️ Engineering Details

_No response_

### PR DESCRIPTION
**Pull request by collinss-jpl**
## 🗒️ Summary
This PR enhances the DUM service to include several user-defined metadata fields with the pre-signed URLs returned to the client. These user-defined fields become associated with the uploaded file once in S3, and can be viewed from the console:

Screenshot 2024-07-23 at 2 43 41 PM

This branch also adds inclusion of the MD5 digest of an uploaded file with the pre-signed URL, so that AWS can validate the integrity of uploaded files automatically at time of ingest to S3.

## ⚙️ Test Data and/or Report
Unit test suite has been updated to accommodate changes to the request payload structure.
New features have been tested in MCP dev environment to verify user-defined metadata is now associated to uploaded S3 objects.

## ♻️ Related Issues

Resolves #87 
Resolves #50

### CODE REVIEW

**Comment by collinss-jpl** [comment]
@jordanpadams Tagging you on this so you can confirm that the screenshot provided above contains all the user-defined metadata you are (currently) looking for.

File: src/pds/ingress/client/pds_ingress_client.py [hunk 1]
```diff
@@ -109,7 +111,10 @@ def perform_ingress(batched_ingress_paths, node_id, prefix, force_overwrite, api
         # Perform uploads to S3 in parallel based on number of files
         PARALLEL(delayed(ingress_file_to_s3)(ingress_response) for ingress_response in chain(*response_batches))
     except Exception as err:
-        logger.error("Ingress failed, reason: %s", str(err))
+        if isinstance(err, RequestException):
```
Comment (nutjob4life) [inline src/pds/ingress/client/pds_ingress_client.py:114]: The rules for exception handling (processed in-order) let you avoid the `isinstance` call:
```python
except RequestException as err:
    logger.error("Ingress failed, HTTP response text:\n%s", err.response.text)
except Exception as err:
    logger.error("Ingress failed, reason: %s", str(err))
```

**Review by nutjob4life — CHANGES_REQUESTED** [review]

**Comment by jordanpadams** [comment]
@collinss-jpl just to confirm, the last modified date is from the file on the file system prior to upload, correct?

File: src/pds/ingress/client/pds_ingress_client.py [hunk 2]
```diff
@@ -214,7 +219,9 @@ def prepare_batch_for_ingress(ingress_path_batch, prefix, batch_index):
 
         # Calculate the MD5 checksum of the file payload
         with open(ingress_path, "rb") as object_file:
-            md5_digest = hashlib.md5(object_file.read()).hexdigest()
```
Comment (nutjob4life) [inline src/pds/ingress/client/pds_ingress_client.py:217]: Weird, my comment on this completely disappeared.

Anyway, to re-iterate: if the file at `ingress_path` is large, then `object_file.read()` will bring the _entire thing_ into memory, which could be a problem. Better to stream it in chunks:
```python
md5 = hashlib.md5()
with open(ingress_path, "rb") as object_file:
    while chunk := object_file.read(4096):
        md5.update(chunk)
md5_digest.hexdigest()
```

**Review by nutjob4life — CHANGES_REQUESTED** [review]

File: src/pds/ingress/client/pds_ingress_client.py (hunk 1 above)
Comment (collinss-jpl) [reply src/pds/ingress/client/pds_ingress_client.py:114]: 🤦 will fix

**Comment by collinss-jpl** [comment]
> @collinss-jpl just to confirm, the last modified date is from the file on the file system prior to upload, correct?

That is correct. Perhaps the name should be something like `local_last_modified` so as not be confused with the S3 last modified time, which AWS manages itself

File: src/pds/ingress/client/pds_ingress_client.py (hunk 2 above)
Comment (collinss-jpl) [reply src/pds/ingress/client/pds_ingress_client.py:217]: Great suggestion, will add

**Review by nutjob4life — APPROVED** [review]

**Comment by nutjob4life** [comment]
@collinss-jpl looks great! Thanks!

**Pull request merged**